src/librustc_unicode/u_str.rs

```diff
@@ -413,67 +412,82 @@ pub fn is_utf16(v: &[u16]) -> bool {
     }
 }
 
-/// An iterator that decodes UTF-16 encoded codepoints from a vector
-/// of `u16`s.
+/// An iterator that decodes UTF-16 encoded codepoints from an iterator of `u16`s.
 #[derive(Clone)]
-pub struct Utf16Items<'a> {
-    iter: slice::Iter<'a, u16>
-}
-/// The possibilities for values decoded from a `u16` stream.
-#[derive(Copy, PartialEq, Eq, Clone, Debug)]
-pub enum Utf16Item {
-    /// A valid codepoint.
-    ScalarValue(char),
-    /// An invalid surrogate without its pair.
-    LoneSurrogate(u16)
+pub struct Utf16Decoder<I> where I: Iterator<Item=u16> {
+    iter: I,
+    buf: u16,
 }
 
-impl Utf16Item {
-    /// Convert `self` to a `char`, taking `LoneSurrogate`s to the
-    /// replacement character (U+FFFD).
-    #[inline]
-    pub fn to_char_lossy(&self) -> char {
-        match *self {
-            Utf16Item::ScalarValue(c) => c,
-            Utf16Item::LoneSurrogate(_) => '\u{FFFD}'
+impl<I: Iterator<Item=u16>> Utf16Decoder<I> {
+    /// Create an iterator over the UTF-16 encoded codepoints in `iter`,
+    /// returning unpaired surrogates as `Err`s.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(unicode)]
+    ///
+    /// extern crate rustc_unicode;
+    /// use rustc_unicode::str::Utf16Decoder;
+    ///
+    /// fn main() {
+    ///     // 𝄞mus<invalid>ic<invalid>
```

Comment:
(I’m just moving this comment, it was previously on the `utf16_items` function: http://doc.rust-lang.org/rustc_unicode/str/fn.utf16_items.html)